File: typescript/tests/metagrammar/emptyParentheses.test.ts
import { MetaGrammar, Parser } from '../../src/index.js';

describe('MetaGrammar - Empty Parentheses (Nothing)', () => {
  test('empty parentheses matches empty string', () => {
    const grammar = `
      Empty <- ();
    `;

    const rules = MetaGrammar.parseGrammar(grammar);
    expect(rules.has('Empty')).toBe(true);

    const parser = new Parser({ topRuleName: 'Empty', rules, input: '' });
    const parseResult = parser.parse();
    const result = parseResult.root;
    expect(result).not.toBeNull();
    expect(result.len).toBe(0);
  });

  test('empty parentheses in sequence', () => {
    const grammar = `
      AB <- "a" () "b";
    `;

    const rules = MetaGrammar.parseGrammar(grammar);
    const parser = new Parser({ topRuleName: 'AB', rules, input: 'ab' });
    const parseResult = parser.parse();
    const result = parseResult.root;
    expect(result).not.toBeNull();
    expect(result.len).toBe(2);
  });

  test('parenthesized expression with content', () => {
    const grammar = `
      Parens <- ("hello");
    `;

    const rules = MetaGrammar.parseGrammar(grammar);
    const parser = new Parser({ topRuleName: 'Parens', rules, input: 'hello' });
    const parseResult = parser.parse();
    const result = parseResult.root;
    expect(result).not.toBeNull();
    expect(result.len).toBe(5);
  });

  test('nested empty parentheses', () => {
    const grammar = `
      Nested <- (());
    `;

    const rules = MetaGrammar.parseGrammar(grammar);
    const parser = new Parser({ topRuleName: 'Nested', rules, input: '' });
    const parseResult = parser.parse();
    const result = parseResult.root;
    expect(result).not.toBeNull();
    expect(result.len).toBe(0);
  });

  test('empty parentheses with optional repetition', () => {
    const grammar = `
      Opt <- ()* "test";
    `;

    const rules = MetaGrammar.parseGrammar(grammar);
    const parser = new Parser({ topRuleName: 'Opt', rules, input: 'test' });
    const parseResult = parser.parse();
    const result = parseResult.root;
    expect(result).not.toBeNull();
    expect(result.len).toBe(4);
  });

  test('empty parentheses in choice', () => {
    const grammar = `
      Choice <- "a" / ();
    `;

    const rules = MetaGrammar.parseGrammar(grammar);

    // Should match 'a'
    const parser1 = new Parser({ topRuleName: 'Choice', rules, input: 'a' });
    const parseResult1 = parser1.parse();
    const result1 = parseResult1.root;
    expect(result1).not.toBeNull();
    expect(result1.len).toBe(1);

    // Should match empty string
    const parser2 = new Parser({ topRuleName: 'Choice', rules, input: '' });
    const parseResult2 = parser2.parse();
    const result2 = parseResult2.root;
    expect(result2).not.toBeNull();
    expect(result2.len).toBe(0);
  });

  test('rule referencing nothing', () => {
    const grammar = `
      Nothing <- ();
      A <- Nothing "a";
    `;

    const rules = MetaGrammar.parseGrammar(grammar);
    const parser = new Parser({ topRuleName: 'A', rules, input: 'a' });
    const parseResult = parser.parse();
    const result = parseResult.root;
    expect(result).not.toBeNull();
    expect(result.len).toBe(1);
  });
});

File: dart/test/metagrammar/empty_parentheses_test.dart
import 'package:squirrel_parser/squirrel_parser.dart';
import 'package:test/test.dart';

void main() {
  group('MetaGrammar - Empty Parentheses (Nothing)', () {
    test('empty parentheses matches empty string', () {
      const grammar = '''
        Empty <- ();
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);
      expect(rules.containsKey('Empty'), isTrue);

      final parser = Parser(topRuleName: 'Empty', rules: rules, input: '');
      var parseResult = parser.parse();
      final result = parseResult.root;
      expect(result, isNotNull);
      expect(result.len, equals(0));
    });

    test('empty parentheses in sequence', () {
      const grammar = '''
        AB <- "a" () "b";
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);
      final parser = Parser(topRuleName: 'AB', rules: rules, input: 'ab');
      var parseResult = parser.parse();
      final result = parseResult.root;
      expect(result, isNotNull);
      expect(result.len, equals(2));
    });

    test('parenthesized expression with content', () {
      const grammar = '''
        Parens <- ("hello");
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);
      final parser = Parser(topRuleName: 'Parens', rules: rules, input: 'hello');
      var parseResult = parser.parse();
      final result = parseResult.root;
      expect(result, isNotNull);
      expect(result.len, equals(5));
    });

    test('nested empty parentheses', () {
      const grammar = '''
        Nested <- (());
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);
      final parser = Parser(topRuleName: 'Nested', rules: rules, input: '');
      var parseResult = parser.parse();
      final result = parseResult.root;
      expect(result, isNotNull);
      expect(result.len, equals(0));
    });

    test('empty parentheses with optional repetition', () {
      const grammar = '''
        Opt <- ()* "test";
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);
      final parser = Parser(topRuleName: 'Opt', rules: rules, input: 'test');
      var parseResult = parser.parse();
      final result = parseResult.root;
      expect(result, isNotNull);
      expect(result.len, equals(4));
    });

    test('empty parentheses in choice', () {
      const grammar = '''
        Choice <- "a" / ();
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);

      // Should match 'a'
      final parser1 = Parser(topRuleName: 'Choice', rules: rules, input: 'a');
      final parseResult1 = parser1.parse();
      final result1 = parseResult1.root;
      expect(result1, isNotNull);
      expect(result1.len, equals(1));

      // Should match empty string
      final parser2 = Parser(topRuleName: 'Choice', rules: rules, input: '');
      final parseResult2 = parser2.parse();
      final result2 = parseResult2.root;
      expect(result2, isNotNull);
      expect(result2.len, equals(0));
    });

    test('rule referencing nothing', () {
      const grammar = '''
        Nothing <- ();
        A <- Nothing "a";
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);
      final parser = Parser(topRuleName: 'A', rules: rules, input: 'a');
      var parseResult = parser.parse();
      final result = parseResult.root;
      expect(result, isNotNull);
      expect(result.len, equals(1));
    });
  });
}
